controller-server/src/main/java/com/yahoo/vespa/hosted/controller/versions/VersionStatus.java

```diff
@@ -202,13 +201,24 @@ private static NodeVersions findSystemApplicationVersions(Controller controller)
                     versionMap.computeIfAbsent(deployment.version(), DeploymentStatistics::empty);
 
             status.instanceJobs().forEach((id, jobs) -> {
-                jobs.failing()
-                    .not().failingApplicationChange()
-                    .not().withStatus(RunStatus.outOfCapacity)
-                    .lastCompleted().mapToList(run -> run.versions().targetPlatform())
-                    .forEach(version -> versionMap.put(version,
-                                                       versionMap.getOrDefault(version, DeploymentStatistics.empty(version))
-                                                                 .withFailing(id)));
+                // Add all unsuccessful runs for failing jobs as any run may have resulted in an incomplete deployment
+                // where a subset of nodes have upgraded.
+                var runsOfFailingJobs = jobs.failing()
+                                            .not().failingApplicationChange()
+                                            .not().withStatus(RunStatus.outOfCapacity).asList().stream()
+                                            .flatMap(job -> job.runs().values().stream())
+                                            .collect(Collectors.toList());
+                var consecutiveFailingRuns = new ArrayList<Run>();
+                for (var run : runsOfFailingJobs) {
+                    if (run.status() == RunStatus.success) {
+                        consecutiveFailingRuns.clear();
+                    } else {
+                        consecutiveFailingRuns.add(run);
+                    }
+                }
+                consecutiveFailingRuns.stream().map(run -> run.versions().targetPlatform())
+                                      .forEach(version -> versionMap.put(version, versionMap.getOrDefault(version, DeploymentStatistics.empty(version))
+                                                                                            .withFailing(id)));
```

Comment:
How about
```java
                jobs.failing()
                    .not().failingApplicationChange()
                    .not().withStatus(RunStatus.outOfCapacity)
                    .mapToList(JobStatus::runs)
                    .forEach(runs -> runs.descendingMap().values().stream()
                                         .dropWhile(run -> ! run.hasEnded())
                                         .takeWhile(run -> run.hasFailed())
                                         .map(run -> run.versions().targetPlatform())
                                         .forEach(version -> versionMap.put(version,
                                                                            versionMap.getOrDefault(version, DeploymentStatistics.empty(version))
                                                                                      .withFailing(id))));
```

the jobs contains runs for all job types, each of which may be relevant — not just the consecutive failures for the last job type. 

Also, jobs which are running (not `success`) aren't failures.

Reply:
Thanks, I got lost in the *List jungle when I tried a different approach. 😅

Reply:
Haha, I know. I get lost myself >_<